### ISSUE TITLE
Chapter 12: how to load Keras model ? 

### ISSUE THREAD

**Issue by avialfont** [foxbook/atap#88]
Hi,

Chapter 12 allows to very simply implement a fitted sklearn Pipeline saved in a .pkl file when the last step is a sklearn nn. For example, after a Pipeline = [TextNormalizer(), TfidfVectorizer(), MLPClassifier()] that is instantiated with the ` train_model()` method saving in a file named 'ann_cls.pkl', the fitted model can then be loaded with : 

```
path='ann_cls.pkl'

with open(path, 'rb') as f:

      model=joblib.load(f)

```

Then, it is possible to predict the class of a new preprocessed element : `model.predict(elem)`. It works perfect, thanks !

However, when the Pipeline ends with Keras method KerasClassifier(), a workaround is proposed, consisting in saving the first steps of the Pipeline in a .pkl file and the Keras object in a .h5 file. Doing the above immediatly implies an AttributeError : 'TfidfVectorizer' object has no attribute 'predict' as the previous step is Tfdif in this exemple.

How can be loaded the Keras nn saved in the .h5 file and then applied after the (two steps) Pipeline stored in the .pkl file ?

Thank you in advance for any help !

**Comment by avialfont** [comment]
I think I have figured it out, as it was quite obvious for those knowing Keras!

Following the save method in Keras, a load method has to be imported to access the .h5 file: 

```
from keras.models import load_model

classifier = load_model('keras_nn.h5') 
```

Then, it is necessary to append the loaded two first stages of the pipeline:
```

path='ann_kcls.pkl'
with open(path, 'rb') as f:
    model=joblib.load(f)

top_model.steps.append(('ann',classifier))
```
Using `model.predict(elem)` then works perfectly, even though it returns an array of probabilities by class instead of the most probable one with the sklearn  MLPClassifier.

Thanks again for this very insteresting and complete book !

**Comment by rebeccabilbro** [comment]
Hello @avialfont and thank you so much for checking out our book! My apologies for not responding sooner to your message from June, it must have slipped through the cracks at the time, but I'm very glad you found a solution in the interim! Thanks again for reading our book and for the kind words; I'll go ahead and archive this issue now in case another reader faces something similar in the future.

**Comment by avialfont** [comment]
Thanks a lot !
    Le vendredi 13 septembre 2019 à 17:14:27 UTC+2, Rebecca Bilbro <notifications@github.com> a écrit :  
 
 
Closed #88.
